spec/ysh-proc.test.sh

```diff
@@ -524,3 +524,35 @@ grep
 ## STDOUT:
 sh-func grep
 ## END
+
+#### proc resolution changes with the local scope
+shopt -s ysh:upgrade
+
+proc foo {
+  echo foo
+}
+
+proc bar {
+  echo bar
+}
+
+proc inner {
+  var foo = bar
+  foo  # Will now reference `proc bar`
+}
+
+foo
+inner
+foo  # Back to the global scope, foo still references `proc foo`
+
+# Without this behavior, features like `eval(b, vars={ flag: __flag })`, needed
```

Comment:
Hm actually I think we may need something more like `ctx_Shvar`

https://github.com/oils-for-unix/oils/blob/master/builtin/pure_ysh.py#L32

It does NOT push a new stack frame, like we are doing here.

Instead it temporarily replaces variable of the given names on SAME stack frame.

The reason we do it that way is because procs don't have dynamic scope.  The problem with pushing a new stack frame is that you will "lose" your current local variables ...  You will ONLY have the new ones.

I think it may come up in `Str => replace()` too -- are the current variables in the frame available for replacement?  It could go either way, but I think for `eval()` for `lines` and `split-by` etc. you will want to be able to use OTHER variables in the same stack frame.

(This was something I learned the hard way -- originally I had pushed a new stack frame, but then that broke some things)

---

Let me see if I can come up with a test case for this ... I will merge this, but then I think we probably will run into this issue, and can revise it

Reply:
Actually, in this case, isn't this good? Previously you could:

```
proc foo(;;; block) {
  var bar = 123
  eval (block)
}

foo { = bar }
# => 123
```

We would leak variables from the `eval` frame into the inner block. With `vars`, you can control which variables are visible (although, you won't get mutation.)

Reply:
The `ctx_Shvar` will save AND restore

But I tried to reproduce the issue I was talking about, and couldn't

https://github.com/oils-for-unix/oils/commit/59ece78259d0eeed24d1f7b51aeec952c365becd

I guess we can move on now

If we hit it, we can fix it then ... I'm pretty sure it's an issue but I wasn't able to come up with an example, and then I got distracted trying to print the stack in a readable format!  That could be useful for tests